File: Problems/Trees/Construct-Binary-Tree-from-Preorder-and-Inorder-Traversal/Construct-Binary-Tree-from-Preorder-and-Inorder-Traversal.cpp
#include <iostream>
#include <vector>
#include <queue>

// Using preorder and inorder traversal - Time: O(n)

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    TreeNode *buildTree(std::vector<int> &preorder, std::vector<int> &inorder)
    {
        if (preorder.empty() || inorder.empty())
        {
            return nullptr;
        }
        return buildTree(preorder, 0, preorder.size(), inorder, 0, inorder.size());
    }

private:
    TreeNode *buildTree(std::vector<int> &preorder, int preorderStart, int preorderEnd, std::vector<int> &inorder, int inorderStart, int inorderEnd)
    {
        TreeNode *root = new TreeNode(preorder[preorderStart]);

        int inorderIndex = inorderStart;
        for (; inorderIndex < inorderEnd; inorderIndex++)
        {
            if (inorder[inorderIndex] == preorder[preorderStart])
            {
                break;
            }
        }

        int leftLength = inorderIndex - inorderStart;

        if (leftLength > 0)
        {
            root->left = buildTree(preorder, preorderStart + 1, preorderStart + 1 + leftLength, inorder, inorderStart, inorderIndex);
        }
        if (inorderEnd > inorderIndex + 1)
        {
            root->right = buildTree(preorder, preorderStart + 1 + leftLength, preorderEnd, inorder, inorderIndex + 1, inorderEnd);
        }

        return root;
    }
};

void printTree(TreeNode *root)
{
    if (!root)
    {
        std::cout << "[]" << std::endl;
        return;
    }

    std::queue<TreeNode *> queue;
    queue.push(root);

    std::cout << "[";

    while (!queue.empty())
    {
        int n = queue.size();

        for (int i = 0; i < n; i++)
        {
            TreeNode *node = queue.front();
            queue.pop();

            if (node)
            {
                std::cout << node->val << ", ";

                queue.push(node->left);
                queue.push(node->right);
            }
            else
            {
                if (!queue.empty())
                    std::cout << "null, ";
                else
                    std::cout << "null";
            }
        }
    }

    std::cout << "]";
}

int main()
{
    std::vector<int> preorder = {3, 9, 20, 15, 7};
    std::vector<int> inorder = {9, 3, 15, 20, 7};

    std::cout << "Input: preorder = [";
    for (int i = 0; i < preorder.size(); i++)
    {
        std::cout << preorder[i];
        if (i < preorder.size() - 1)
        {
            std::cout << ", ";
        }
    }
    std::cout << "], inorder = [";
    for (int i = 0; i < inorder.size(); i++)
    {
        std::cout << inorder[i];
        if (i < inorder.size() - 1)
        {
            std::cout << ", ";
        }
    }
    std::cout << "]" << std::endl;

    Solution solution;
    TreeNode *root = solution.buildTree(preorder, inorder);

    std::cout << "Output: ";
    printTree(root);
    std::cout << std::endl;

    delete root;

    return 0;
}

File: Problems/Trees/Construct-Binary-Tree-from-Preorder-and-Inorder-Traversal/Construct-Binary-Tree-from-Preorder-and-Inorder-Traversal.cs
using System;

// Using preorder and inorder traversal - Time: O(n)

public class TreeNode
{
    public int val;
    public TreeNode left;
    public TreeNode right;

    public TreeNode(int val = 0, TreeNode left = null, TreeNode right = null)
    {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

public class Solution
{
    public TreeNode BuildTree(int[] preorder, int[] inorder)
    {
        if (preorder.Length == 0 || inorder.Length == 0) 
        {
            return null;
        }
        return BuildTree(preorder, 0, preorder.Length, inorder, 0, inorder.Length);
    }

    private TreeNode BuildTree(int[] preorder, int preorderStart, int preorderEnd, int[] inorder, int inorderStart, int inorderEnd)
    {
        var root = new TreeNode(preorder[preorderStart]);

        var inorderIndex = inorderStart;
        for (; inorderIndex < inorderEnd; inorderIndex++)
        {
            if (inorder[inorderIndex] == preorder[preorderStart])
            {
                break;
            }
        }


        var leftLength = inorderIndex - inorderStart;

        if (leftLength > 0)
        {
            root.left = BuildTree(preorder, preorderStart + 1, preorderStart + 1 + leftLength, inorder, inorderStart, inorderIndex);
        }
        if (inorderEnd > inorderIndex + 1)
        {
            root.right = BuildTree(preorder, preorderStart + 1 + leftLength, preorderEnd, inorder, inorderIndex + 1, inorderEnd);
        }
        
        return root;
    }
}

class Program
{
    static void PrintTree(TreeNode root)
    {
        if (root == null)
        {
            Console.Write("null");
            return;
        }

        List<string> values = new List<string>();
        Queue<TreeNode> queue = new Queue<TreeNode>();
        queue.Enqueue(root);

        while (queue.Count > 0)
        {
            TreeNode current = queue.Dequeue();
            if (current != null)
            {
                values.Add(current.val.ToString());
                queue.Enqueue(current.left);
                queue.Enqueue(current.right);
            }
            else
            {
                values.Add("null");
            }
        }

        Console.Write("[" + string.Join(", ", values) + "]");
    }
    static void Main()
    {
        int[] preorder = { 3, 9, 20, 15, 7 };
        int[] inorder = { 9, 3, 15, 20, 7 };

        Console.WriteLine("Input: preorder = [" + string.Join(", ", preorder) + "], inorder = [" + string.Join(", ", inorder) + "]");

        Solution sol = new Solution();
        TreeNode root = sol.BuildTree(preorder, inorder);

        Console.Write("Ouput: ");
        PrintTree(root);
        Console.WriteLine();
    }
}
